torch/jit/__init__.py

```diff
@@ -456,7 +457,152 @@ def trace(*args, **kwargs):
         >>> def f(x):
         >>>     return x * 2
     """
-    return lambda func: torch._C.GraphExecutor(func, args, kwargs.pop('optimize', True))
+    def wrapper(func):
+        executor_options = {'optimize': True}
+        for name in executor_options:
+            executor_options[name] = kwargs.pop(name, executor_options[name])
+        if isinstance(func, torch.nn.Module):
+            captures = list(func.state_dict(keep_vars=True).values())
+            # TODO: support shared parameters
+            if len(set(map(id, captures))) != len(list(map(id, captures))):
+                raise ValueError("TracedModules don't support parameter sharing between modules")
+            executor = torch._C.GraphExecutor(func, args, captures=captures, **executor_options)
+            return TracedModule(func, executor)
+        else:
+            return torch._C.GraphExecutor(func, args, **executor_options)
+    return wrapper
+
+
+class TracedModuleBase(torch.nn.Module):
+    __frozen = False
+
+    def __init__(self, orig):
+        super(TracedModuleBase, self).__init__()
+
+        self.training = orig.training
+        for name, param in orig._parameters.items():
+            if param is not None:
+                self._parameters[name] = param
+        for name, buf in orig._buffers.items():
+            if param is not None:
+                self._buffers[name] = buf
+        self._orig_class = type(orig)
+
+        if orig._backward_hooks or orig._forward_hooks or orig._forward_pre_hooks:
+            raise ValueError("Modules that have hooks assigned can't be compiled")
+
+        # XXX: submodules can't be initialized here, because we don't know what the root is
+
+    def _freeze(self):
+        self.__frozen = True
+
+    def __setattr__(self, name, value):
+        if not self.__frozen:
+            return super(TracedModuleBase, self).__setattr__(name, value)
+        if name in self._parameters or name in self._buffers:
+            self._recompute_captures()
+            return super(TracedModuleBase, self).__setattr__(name, value)
```

Comment:
Shouldn't `_recompute_captures()` go after the super call to `__setattr__`?

Reply:
Yeah I think it should.